File: apps/mobile/vet_app/lib/core/api_config.dart
import 'package:shared_preferences/shared_preferences.dart';

import '../config/app_config.dart';

/// Central API config — single source of truth. Change here only.
/// Supports: LAN / 127.0.0.1 or ngrok URL (https://xxx.ngrok-free.app)
class ApiConfig {
  static const _keyHost = 'api_host_override';

  static String _defaultHost() => AppConfig.defaultHostValue();

  static Future<String> getHost() async {
    final prefs = await SharedPreferences.getInstance();
    final override = prefs.getString(_keyHost);
    return override?.trim().isEmpty ?? true ? _defaultHost() : override!.trim();
  }

  static bool _isFullUrl(String value) =>
      value.startsWith('http://') || value.startsWith('https://');

  static Future<String> getBaseUrl() async {
    final value = await getHost();
    if (_isFullUrl(value)) {
      final base =
          value.endsWith('/') ? value.substring(0, value.length - 1) : value;
      return '$base/api/v1';
    }
    return 'http://$value:3000/api/v1';
  }

  /// Socket.io server origin (same as API host, no path).
  static Future<String> getSocketUrl() async {
    final value = await getHost();
    if (_isFullUrl(value)) {
      return value.endsWith('/') ? value.substring(0, value.length - 1) : value;
    }
    return 'http://$value:3000';
  }

  static Future<void> setHost(String host) async {
    final prefs = await SharedPreferences.getInstance();
    final trimmed = host.trim();
    if (trimmed.isEmpty) {
      await prefs.remove(_keyHost);
    } else {
      await prefs.setString(_keyHost, trimmed);
    }
  }

  static Future<void> clearOverride() async {
    await setHost('');
  }

  static Future<bool> hasOverride() async {
    final prefs = await SharedPreferences.getInstance();
    final v = prefs.getString(_keyHost);
    return v != null && v.trim().isNotEmpty;
  }

  static Map<String, String> ngrokHeadersForBaseUrl([String? _]) {
    return {'ngrok-skip-browser-warning': 'true'};
  }
}

File: apps/web/admin/lib/apiConfig.ts
/**
 * Single source of truth for admin → Node API wiring.
 * Backend default port is 3000; REST prefix is /api/v1 (not :5000, not /api alone).
 *
 * Set NEXT_PUBLIC_API_URL to your stable backend (e.g. ngrok https://….ngrok-free.app/api/v1).
 * For local-only dev without NEXT_PUBLIC_API_URL, set NEXT_PUBLIC_DEV_API_BASE_URL and
 * optionally NEXT_PUBLIC_DEV_SOCKET_URL (no hardcoded localhost in source).
 */
function trimSlash(s: string) {
  return s.replace(/\/+$/, '');
}

function devApiFallback(): string {
  return (process.env.NEXT_PUBLIC_DEV_API_BASE_URL || '').trim();
}

function devSocketFallback(): string {
  return (process.env.NEXT_PUBLIC_DEV_SOCKET_URL || '').trim();
}

function originFromApiBase(base: string): string {
  const raw = base.trim();
  if (!raw) return '';
  try {
    const u = new URL(raw.startsWith('http') ? raw : `https://${raw}`);
    return `${u.protocol}//${u.host}`;
  } catch {
    return '';
  }
}

/**
 * REST base URL, e.g. https://xxx.ngrok-free.app/api/v1
 * If NEXT_PUBLIC_API_URL is an ngrok URL, the browser uses same-origin `/api/v1` and Next.js
 * rewrites proxy to that tunnel (avoids CORS preflight hitting ngrok without bypass header).
 */
export function getAdminApiBaseUrl(): string {
  const raw = process.env.NEXT_PUBLIC_API_URL?.trim();
  if (raw) {
    if (/ngrok/i.test(raw)) {
      return '/api/v1';
    }
    return trimSlash(raw);
  }
  if (process.env.NODE_ENV === 'development') {
    const fb = devApiFallback();
    if (fb) return trimSlash(fb);
  }
  return '';
}

/**
 * Always send ngrok bypass on outbound axios calls (harmless when not using ngrok).
 * Preflight OPTIONS from the browser cannot attach this header; same-origin API proxy covers that.
 */
export function getNgrokBrowserBypassHeaders(): Record<string, string> {
  return { 'ngrok-skip-browser-warning': 'true' };
}

/**
 * Socket.io origin (no /api/v1 path). When REST is ngrok-proxied via Next, connect sockets to the
 * real Node server unless NEXT_PUBLIC_SOCKET_URL or dev fallbacks override — avoids browser→ngrok socket preflight.
 */
/**
 * True when env points at loopback or no API URL is configured (helps tunnel / public-host hints).
 */
export function adminApiEnvPointsLocalOrUnset(): boolean {
  const raw = process.env.NEXT_PUBLIC_API_URL?.trim() || '';
  const dev = process.env.NEXT_PUBLIC_DEV_API_BASE_URL?.trim() || '';
  const primary = raw || (process.env.NODE_ENV === 'development' ? dev : '');
  if (!primary) return true;
  const lower = primary.toLowerCase();
  return lower.includes('localhost') || lower.includes('127.0.0.1');
}

export function getAdminSocketUrl(): string {
  const socketEnv = process.env.NEXT_PUBLIC_SOCKET_URL?.trim();
  if (socketEnv) return trimSlash(socketEnv);

  const api = process.env.NEXT_PUBLIC_API_URL?.trim();
  if (api && /ngrok/i.test(api)) {
    return trimSlash(devSocketFallback() || originFromApiBase(devApiFallback()) || '');
  }
  if (api) {
    try {
      const u = new URL(api);
      return `${u.protocol}//${u.host}`;
    } catch {
      /* ignore */
    }
  }
  if (process.env.NODE_ENV === 'development') {
    const ds = devSocketFallback();
    if (ds) return trimSlash(ds);
    const da = devApiFallback();
    const fromDa = originFromApiBase(da);
    if (fromDa) return fromDa;
  }
  return '';
}
